[BulkSimd PR1] Added MMapByteVectorValues for FP16 in LuceneOnFaiss.

0ctopus13prime opened the pull request:
### Description
This is the first PR to support native scoring in LuceneOnFaiss. 

RFC : https://github.com/opensearch-project/k-NN/issues/2875

In this first PR, FP16 flat index will return `MMapByteVectorValues` having `MemorySegment[]` extracted from MemorySegment$IndexInput. Each `MemorySegment` will have a raw pointer pointing to mapped region in memory.
Passing the raw pointer for distance calculation will be dealt with in the second PR. This PR is mostly set-up for the second PR.

Since the actual distance calculation is not there, still the distance calculation will be conducted in Java side.
This PR is mostly focusing on extracting `MemorySegment` from IndexInput.


### Related Issues
Resolves #[Issue number to be closed when this PR is merged]
N/A

### Check List
- [O] New functionality includes testing.
- [O] New functionality has been documented.
- [O] API changes companion pull request [created](https://github.com/opensearch-project/opensearch-api-specification/blob/main/DEVELOPER_GUIDE.md).
- [O] Commits are signed per the DCO using `--signoff`.
- [O] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose).

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/k-NN/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

0ctopus13prime commented on src/main/java/org/opensearch/knn/memoryoptsearch/MemorySegmentAddressExtractorJDK22.java:
```diff
@@ -0,0 +1,57 @@
+            }
+            return addressAndSize;
+        } catch (NoSuchFieldException | IllegalAccessException ea) {
+            // Ignore
```
Will add it in next rev.

0ctopus13prime commented on src/main/java/org/opensearch/knn/memoryoptsearch/MemorySegmentAddressExtractorJDK21.java:
```diff
@@ -0,0 +1,87 @@
+     * @return an array of address and size info extracted from the input, or null if it's not found.
+     *         ex: address_i = array[i], size_i = array[i + 1]
+     */
+    public long[] extractAddressAndSize(final IndexInput indexInput) {
```
Will add it in next rev.

0ctopus13prime commented on src/main/java/org/opensearch/knn/memoryoptsearch/MemorySegmentAddressExtractorJDK21.java:
```diff
@@ -0,0 +1,87 @@
+            }
+
+            return addressAndSize;
+        } catch (NoSuchFieldException | IllegalAccessException | IllegalArgumentException e) {
+            // Ignore
```
Sure, will add it in next rev.

navneet1v commented on src/main/java/org/opensearch/knn/memoryoptsearch/faiss/MMapByteVectorValues.java:
```diff
@@ -0,0 +1,73 @@
+        this.addressAndSize = addressAndSize;
+    }
+
+    @Override
+    public byte[] vectorValue(int internalVectorId) throws IOException {
+        indexInput.seek(baseOffset + internalVectorId * oneVectorByteSize);
+        // Lazy initialization, in general this method is not expected to be called during search.
+        // During search, distance calculation will be done in a separated C++ code.
+        if (buffer == null) {
+            buffer = new byte[(int) oneVectorByteSize];
+        }
+        indexInput.readBytes(buffer, 0, buffer.length);
+        return buffer;
+    }
```
So here what I would suggest, please add java doc and be little verbose in explaining what this function is doing and explanation of parameters. This will help in future.

navneet1v commented on src/main/java/org/opensearch/knn/memoryoptsearch/MemorySegmentAddressExtractorUtil.java:
```diff
@@ -0,0 +1,32 @@
+            instance = (MemorySegmentAddressExtractor) clazz.getDeclaredConstructor().newInstance();
+        } catch (ClassNotFoundException e) {
+            // Class not found: fall back to JDK21 version
+            instance = new MemorySegmentAddressExtractorJDK21();
+        } catch (Throwable t) {
+            // Any other errors (constructor, reflection issues)
+            throw new RuntimeException("Failed to initialize MemorySegmentAddressExtractor", t);
+        }
```
@0ctopus13prime we might need to think little bit more here, if you can provide some details why you went with killing of node by design approach that can help.

navneet1v replied:
> Rather than falling back to Java scoring silently, I think it would be better to abort the node.

This can cause a widespread outage even when non k-NN indices are present if there are bugs in this code. I think it will be better if you do lazy init.

0ctopus13prime replied:
Why the whole node will fail? 
The util is being used in FaissIndexScalarQuantizedFlat for now, that means it is trying to do vector search operation on a shard level, if it fails, then vector search will fail and I believe the whole node will remain safe... 🤔 

Any failures within static initialization will throw an exception if it's being captured, then JVM process will keep alive.

But anyway, I've updated to assign empty extractor will make the logic fallback to Java based distance calculation with error logging.

Vikasht34 commented on src/main/java/org/opensearch/knn/memoryoptsearch/MemorySegmentAddressExtractorJDK21.java:
```diff
@@ -0,0 +1,87 @@
+        try {
+            // MMapDirectory in Lucene will return MemorySegmentIndexInput$SingleSegmentImpl or .$MultiSegmentImpl.
+            // Thus, get the super class (e.g. MemorySegmentIndexInput) to acquire `MemorySegment[] segments`.
+            final Field f = indexInput.getClass().getSuperclass().getDeclaredField("segments");
```
this code assumes that ,immediate superclass has a private field names exactly "segments". That's an internal  to lucene and can be changed, if lucene changes it m, this will quietly fail. Do u think we should handle upfront when we are takling such assumption?

0ctopus13prime replied:
But if there's such changes in Lucene, then it probably will be shipped into a newer version, in turn,  those related unit tests will fail anyway. So we have to fix the fixture if there's change in Lucene regardless

0ctopus13prime commented on src/main/java/org/opensearch/knn/memoryoptsearch/faiss/FaissMemoryOptimizedSearcher.java:
```diff
@@ -67,6 +69,16 @@ public FaissMemoryOptimizedSearcher(final IndexInput indexInput, final FieldInfo
+        // Use native scoring logic to calculate distance outside JVM.
+        this.useNativeScorer = false;
+        try {
+            if (faissIndex.getByteValues(indexInput) instanceof MMapByteVectorValues) {
```
Will update in the next rev.

Vikasht34 reviewed (commented):
Hoping to understand more on this , Correct me if I am wrong !!. We are adding reflection on our side,  What it takes to enhance IndexInput on Lucene Side apis to give this instead of doing in KNN. Can we propose supported SPI from Lucene instead of creating in KNN., Instead of ceating mmap vectorValues , What it takes to use FloatVectorValues or Byte from Lcuene.

0ctopus13prime commented on src/main/java/org/opensearch/knn/memoryoptsearch/MemorySegmentAddressExtractorJDK21.java:
```diff
@@ -0,0 +1,87 @@
+            final int numSegments = Array.getLength(objSegments);
+            final long[] addressAndSize = new long[2 * numSegments];
+            for (int i = 0; i < numSegments; i += 2) {
+                final Object memorySegment = Array.get(objSegments, i);
```
Sounds good, will update in the next rev.

0ctopus13prime commented on src/main/java/org/opensearch/knn/memoryoptsearch/MemorySegmentAddressExtractorJDK21.java:
```diff
@@ -0,0 +1,87 @@
+            }
+
+            return addressAndSize;
+        } catch (NoSuchFieldException | IllegalAccessException | IllegalArgumentException e) {
```
Reflection failing indicates either non-MMapDirectory is being used or user intentionally blocked using reflection. 
For those case, nothing we can do about it to enforce using this bulk SIMD optimization. 
I think we can silently fallback to the default scoring which will use Java code, but I think we should log warm at least.
Will update in the next rev.

Vikasht34 commented on src/main/java/org/opensearch/knn/memoryoptsearch/MemorySegmentAddressExtractorJDK21.java:
```diff
@@ -0,0 +1,87 @@
+            }
+            final int numSegments = Array.getLength(objSegments);
+            final long[] addressAndSize = new long[2 * numSegments];
+            for (int i = 0; i < numSegments; i += 2) {
```
Suppose We have 3 segements , with the loop we are just doing 0, 2 and skipped 1. but the suggetion I provided , it does not skip odd number sgements.

0ctopus13prime commented:
> overall looks good. One thing I would suggest is for JDK21 and even JDK22 and above, we should have a jvm args that can switch off the `MemorySegmentAddressExtractor` feature. See in which PR you want to add this capability.

I can add this in the last PR.

0ctopus13prime commented on src/main/java/org/opensearch/knn/memoryoptsearch/MemorySegmentAddressExtractorUtil.java:
```diff
@@ -0,0 +1,78 @@
+                Class<?> clazz = Class.forName("org.opensearch.knn.memoryoptsearch.MemorySegmentAddressExtractorJDK22");
+                instance = (MemorySegmentAddressExtractor) clazz.getDeclaredConstructor().newInstance();
+            } catch (ClassNotFoundException e) {
+                log.warn("Failed to load MemorySegmentAddressExtractorJDK22, falling back to MemorySegmentAddressExtractorJDK21");
```
sure will do.

0ctopus13prime commented:
Hi @Vikasht34 @navneet1v @shatejas 
I think I've addressed all comments, could you take a look and and approve it if all look good on you?
Please note that this is to merge against the feature branch.
Thank you.

Vikasht34 reviewed (approved):
Looks Good , Thanks for addressing the commnets
